File: AllSpice/Services/RecipesService.cs




namespace AllSpice.Services;

public class RecipesService
{
    private readonly RecipesRepository _repo;


    public RecipesService(RecipesRepository repo)
    {
        _repo = repo;
    }

    internal Recipe CreateRecipe(Recipe recipeData)
    {
        Recipe recipe = _repo.CreateRecipe(recipeData);
        return recipe;
    }

    internal Recipe GetRecipeById(int recipeId)
    {
        Recipe recipe = _repo.GetRecipeById(recipeId);
        if (recipe == null) throw new Exception("No recipe with Id " + recipe.Id);
        return recipe;
    }

    internal Recipe EditRecipe(Recipe recipeData)
    {
        Recipe og = this.GetRecipeById(recipeData.Id);
        if (og.CreatorId != recipeData.CreatorId) throw new Exception("You Don't Own This!");
        og.Title = recipeData.Title ?? og.Title;
        og.Instructions = recipeData.Instructions ?? og.Instructions;
        og.Img = recipeData.Img ?? og.Img;
        og.Category = recipeData.Category ?? og.Category;

        Recipe recipe = _repo.EditRecipe(og);
        return recipe;

    }

    internal List<Recipe> GetRecipes()
    {
        List<Recipe> recipes = _repo.GetRecipes();
        return recipes;
    }

    internal void DeleteRecipe(int recipeId, string userId)
    {
        Recipe og = this.GetRecipeById(recipeId);
        if (og.CreatorId != userId) throw new Exception("You Don't Own this!");
        _repo.DeleteRecipe(recipeId);
    }

    // INTERALS
}

File: client/src/services/RecipesService.js
import { AppState } from "../AppState"
import { Recipe } from "../models/Recipe"
import { api } from "./AxiosService"

class RecipesService{

    async createRecipe(recipeData) {
        const res = await api.post('api/recipes', recipeData)
        const newRecipe = new Recipe(res.data)
        AppState.recipes.push(newRecipe)
        AppState.activeRecipe = newRecipe
    }
    async getAllRecipes() {
        const res = await api.get('api/recipes')
        AppState.recipes = res.data.map(recipe => new Recipe(recipe))
        AppState.activeRecipes = AppState.recipes
    }

    async editRecipe(recipe){
        await api.put(`api/recipes/${recipe.id}`, recipe)
    }

    async deleteRecipe(recipeId){
        await api.delete(`api/recipes/${recipeId}`)
        AppState.activeRecipe = {}
        const allfoundIndex = AppState.recipes.findIndex(recipe => recipe.id == recipeId)
        AppState.recipes.splice(allfoundIndex, 1)

        const favoriteIndex = AppState.favorites.findIndex(recipe => recipe.id == recipeId)
        AppState.favorites.splice(favoriteIndex, 1)
        const activefoundIndex = AppState.activeRecipes.findIndex(recipe => recipe.id == recipeId)
        AppState.activeRecipes.splice(activefoundIndex, 1)
    }

    setActiveRecipe(recipeId) {
        AppState.activeRecipe = AppState.recipes.find(recipe => recipe.id == recipeId)
    }

    changeRecipes(recipeType){
        let activeRecipes = AppState.activeRecipes
        const allRecipes = AppState.recipes
        AppState.activeRecipes = []
        if(recipeType == "home")
        {
            activeRecipes = allRecipes
        }
        if(recipeType == "mine")
        {
            activeRecipes = allRecipes.filter(recipe => recipe.creatorId == AppState.account.id)
        }
        AppState.activeRecipes = activeRecipes
    }

    searchRecipes(searchData){
        let nameFiltered = AppState.recipes.filter(recipe => recipe.title.toLowerCase() == searchData.search.toLowerCase())
        let categoryFiltered = AppState.activeRecipes = AppState.recipes.filter(recipe => recipe.category.toLowerCase() == searchData.search.toLowerCase())
        let activeRecipes = [...nameFiltered, ...categoryFiltered]
        AppState.activeRecipes=activeRecipes
    }
}
export const recipesService = new RecipesService()